### PR TITLE
fix(select): Fix typeahead behavior and scrolling

### LINKED ISSUE #755: Selectbox component does not scroll to selected item list
## 🐛 Bug Report

When using a Select box to display a list of items, everything seems to work fine apart from the scolling logic. When making a selection, the menu list doesnt properly scroll to the selected item.
The item list must be > 15 items

## To Reproduce

1. Create a Select box component. Add a list of atleast 20 items.(I added 40 items). 
2. Make a selection. 
3. Open select box again. Expect menu to scroll to selected item.
4. Menu is not able to correctly calculate scroll position

## Expected Behavior

Select box is supposed to scroll when the selected item is not within the visible results of no scroll offset. Select box should properly scroll to selected item

## Actual Results

Select box isnt able to calculate scroll position correctly

## Link to repl or repo (highly encouraged)
https://codesandbox.io/s/dawn-haze-8dnmx 

Please provide a reference to a repo or branch to help us reproduce this bug.
https://github.com/Workday/canvas-kit/blob/master/modules/_labs/select/react/lib/scrolling.ts#L33

```bash

```

or

Include a screenshot of a visual bug

### PR DESCRIPTION
**Pull request by jamesfan**
## Summary

Fixes various issues in the Select component related to assistive focus and scrolling.

1. Fixes an issue where typing multiple characters in rapid succession cycled assistive focus through options incorrectly if there were multiple options that began with the same string. For example, take a Select with two options: `San Francisco` and `San Mateo`.
    * Previous incorrect behavior: typing "san" alternated focus between `San Francisco` and `San Mateo`
    * Corrected behavior: typing "san" maintains focus on `San Francisco`. Note that typing "san m" will maintain focus on `San Francisco` for "san " and then shift focus to `San Mateo` when the "m" is typed at the end.
2. Fixes an issue where focusing an out-of-view option while the menu was open wasn't scrolling to the newly focused option.
3. Fixes an issue where a newly opened menu wasn't scrolling the selected option into view (Fixes #755).
4. Adds Cypress tests for the behaviors specified above as well as various other focus- and scrolling-related behaviors.

### CODE REVIEW

**Comment by lychyi** [comment]
@jamesfan Is this one still technically in draft or is it ready to review?

File: modules/_labs/select/react/lib/SelectBase.tsx [hunk 1]
```diff
@@ -293,17 +293,27 @@ export default class SelectBase extends React.Component<SelectBaseProps> {
...
 
     // If the menu was just displayed, scroll the focused option into
     // center view
     if (!isMenuHidden && prevProps.isMenuHidden) {
-      this.scrollFocusedOptionIntoView(true);
+      // Delay scrolling by a frame to ensure proper measurements of DOM elements
+      // so we know how far to scroll. Without this delay, sometimes measurements
+      // are correct and sometimes they aren't (may be related to the number of
+      // options and/or the length of their labels) -- add the delay to be safe.
+      requestAnimationFrame(() => {
```
Comment (jamesfan) [inline modules/_labs/select/react/lib/SelectBase.tsx:305]: This `requestAnimationFrame` call will be removed when we convert `SelectBase` into a function component and use hooks for the scrolling behavior in #759.

File: cypress/integration/SelectLabs.spec.ts [hunk 2]
```diff
@@ -366,4 +352,284 @@ describe('Select', () => {
       });
     });
   });
+
+  context(`given the "Scrollable" story is rendered`, () => {
```
Comment (jamesfan) [inline cypress/integration/SelectLabs.spec.ts:356]: I moved all of the tests related to typeahead and scrolling to the Scrollable story context for improved readability. The Scrollable story also provides a more complete basis for testing the scrolling-related behaviors of Select given its longer list of options.

File: cypress/integration/SelectLabs.spec.ts [hunk 3]
```diff
@@ -366,4 +352,284 @@ describe('Select', () => {
...
+      h.stories.load('Labs|Select/React/Top Label', 'Scrollable');
+    });
+
+    context('when the select button is focused', () => {
+      beforeEach(() => {
+        cy.findByLabelText('Label').focus();
+      });
+
+      context(
+        'when a character is typed (provided no other characters have been typed in the last 500ms), the select should select the first matching option beyond the currently selected option (cycling back to the beginning of the options if necessary)',
```
Comment (jamesfan) [inline cypress/integration/SelectLabs.spec.ts:367]: I've added detailed descriptions for the typeahead and scrolling tests using `context`. The tests felt a bit random without this additional information -- it seemed unclear why we were testing certain things, whereas this additional context better conveys the intent behind our testing. Let me know if anyone thinks there's a better way to communicate this information.

**Pull request merged**